File: wagerproof-ios-native/Wagerproof/Features/Agents/Components/ToolActivityStack.swift
import SwiftUI
import WagerproofDesign

/// The horizontal fan of placeholder pick tickets shown while polling. Every
/// cumulative tool call the agent makes (`count`) deals one more BLANK scaffold
/// ticket — the exact `AgentPickMiniTicketSkeleton` used by the final Today's
/// Picks rail — into a left-aligned, evenly-fanned deck.
///
/// Design language (matches the reference mock):
///   • **Left-aligned, grows across the screen.** The first ticket sits at the
///     leading edge; each newer ticket fans out to its right. The row extends
///     rightward as tickets deal in, then tightens its overlap once it reaches
///     the trailing edge so it never runs past the width.
///   • **First ticket on top.** The leading (oldest) ticket is frontmost; newer
///     ones fan BEHIND it to the right. So the "beginning" of the run stays fully
///     lit and unoccluded — it never washes out.
///   • **Even coloring.** Every ticket renders at the same brightness/scale.
///     Depth reads purely from the crisp drop-shadow seam each ticket casts onto
///     the one fanned behind it — no dimming or shrinking of the back cards.
struct ToolActivityStack: View {
    /// Cumulative tool-use count from the run metadata (`toolCalls`).
    var count: Int
    /// Kept for call-site compatibility; the skeleton cardstock is neutral.
    var accent: Color

    /// How many scaffolds render at once before the oldest fall off the front.
    private let maxVisible = 7
    /// Comfortable peek while the deck is still growing (few cards) — how much of
    /// each ticket's leading edge shows before the next one fans over it. Shrinks
    /// automatically once the fan reaches the trailing edge.
    private let idealPeek: CGFloat = 96
    /// Never overlap tighter than this, so every ticket keeps a readable sliver.
    private let minPeek: CGFloat = 28

    private var cardW: CGFloat { AgentPickMiniTicket.width }
    private var cardH: CGFloat { AgentPickMiniTicket.height }

    private var visibleIndices: [Int] {
        guard count > 0 else { return [] }
        let lower = max(0, count - maxVisible)
        return Array(lower..<count)
    }

    var body: some View {
        GeometryReader { geo in
            let indices = visibleIndices
            let n = indices.count
            // The trailing ticket's leading edge can travel at most this far and
            // still keep its trailing edge inside the container.
            let travel = max(0, geo.size.width - cardW)
            // Grow at `idealPeek` until the fan reaches the trailing edge, then
            // clamp so it exactly fills the width (tighter overlap), with a floor.
            let fillPeek = n > 1 ? travel / CGFloat(n - 1) : 0
            let peek = n > 1 ? max(minPeek, min(idealPeek, fillPeek)) : 0

            ZStack(alignment: .leading) {
                ForEach(indices, id: \.self) { idx in
                    let pos = idx - (indices.first ?? 0)   // 0 = leading / frontmost
                    AgentPickMiniTicketSkeleton()
                        .offset(x: CGFloat(pos) * peek)
                        // Front-to-back: the leading ticket is on top; each casts
                        // its shadow rightward onto the one fanned behind it → a
                        // crisp depth seam with no dimming or scaling.
                        .zIndex(Double(maxVisible - pos))
                        .shadow(color: .black.opacity(0.55), radius: 9, x: 7, y: 4)
                        .transition(.asymmetric(
                            insertion: .move(edge: .trailing).combined(with: .opacity),
                            removal: .opacity
                        ))
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
            .animation(.spring(response: 0.44, dampingFraction: 0.82), value: count)
        }
        // Reserve a stable footprint (+ a little slack for the shadow) so the row
        // below never jumps as tickets deal in.
        .frame(height: cardH + 10)
    }
}

#if DEBUG
#Preview("Tool activity stack") {
    struct Demo: View {
        @State private var count = 1
        var body: some View {
            ZStack {
                Color(hex: 0x0B1011).ignoresSafeArea()
                VStack(spacing: 24) {
                    ToolActivityStack(count: count, accent: Color(hex: 0xFF8A00))
                    HStack(spacing: 16) {
                        Button("Deal ticket") { count += 1 }
                        Button("Reset") { count = 1 }
                    }
                    .foregroundStyle(.white)
                    Text("count: \(count)").foregroundStyle(.secondary)
                }
                .padding(16)
            }
            .preferredColorScheme(.dark)
        }
    }
    return Demo()
}
#endif

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/components/ToolActivityStack.kt
package com.wagerproof.app.features.agents.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.MutableTransitionState
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInHorizontally
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex
import com.wagerproof.app.features.agents.AgentTicketGeometry

/**
 * Fanned deck of blank pick-ticket scaffolds shown while a generation run polls —
 * port of iOS `ToolActivityStack.swift`. Every cumulative tool call the agent
 * makes deals one more [AgentPickMiniTicketSkeleton] into the fan, so a stalled
 * run visibly stops dealing while a healthy one keeps growing.
 *
 * Layout contract (matches iOS):
 *   - Left-aligned; the first (oldest) ticket sits at the leading edge and is
 *     FRONTMOST, newer ones fan out behind it to the right.
 *   - Grows at [TOOL_STACK_IDEAL_PEEK] until the fan reaches the trailing edge, then tightens
 *     the overlap to exactly fill the width, never below [TOOL_STACK_MIN_PEEK].
 *   - Even coloring: depth reads from stacking order alone, no dimming/scaling.
 */
@Composable
fun ToolActivityStack(
    count: Int,
    /** Kept for call-site compatibility (iOS `ToolActivityStack.accent`) — the skeleton cardstock is neutral. */
    accent: Color = Color.Unspecified,
    modifier: Modifier = Modifier,
) {
    val cardW = AgentTicketGeometry.MINI_WIDTH
    val cardH = AgentTicketGeometry.MINI_HEIGHT
    val targetIndices = toolActivityVisibleIndices(count)
    // Tickets currently mounted, including ones fading OUT after falling off
    // the 7-card window — a plain list-diff would pop them instantly (iOS fades).
    val mounted = remember { mutableStateListOf<Int>() }
    for (idx in targetIndices) if (idx !in mounted) mounted.add(idx)

    // Reserve a stable footprint even at count == 0 so the status line below
    // never jumps as the first ticket deals in.
    BoxWithConstraints(
        modifier = modifier.height(cardH + 10.dp),
        contentAlignment = Alignment.CenterStart,
    ) {
        if (mounted.isEmpty()) return@BoxWithConstraints
        val peek = toolActivityPeek(
            visibleCount = targetIndices.size,
            containerWidth = maxWidth,
            cardWidth = cardW,
        )
        val first = targetIndices.firstOrNull() ?: mounted.min()
        mounted.sorted().forEach { idx ->
            // Absolute tool-call identity matters once the seven-card window
            // starts sliding. Positional nodes made calls 8+ look completely
            // static because every slot retained the same offset state.
            key(idx) {
                // A ticket still finishing its fade-out after falling off the front of the
                // sliding window has idx < first, i.e. a negative pos — clamped to 0 so it
                // settles at the leading edge instead of flying left over the deck with an
                // out-of-range zIndex above every visible card.
                val pos = (idx - first).coerceAtLeast(0) // 0 = leading / frontmost
                // Animate the x-offset so retained tickets visibly make room
                // for the newly dealt trailing card.
                val x by animateDpAsState(
                    targetValue = peek * pos,
                    animationSpec = spring(dampingRatio = 0.82f, stiffness = Spring.StiffnessMediumLow),
                    label = "toolTicketOffset-$idx",
                )
                // Deal-in from the trailing edge + fade; fade-out on removal —
                // port of iOS's `.transition(.asymmetric(insertion: .move(edge:
                // .trailing).combined(with: .opacity), removal: .opacity))`.
                val visibleState = remember { MutableTransitionState(false) }
                visibleState.targetState = idx in targetIndices
                if (!visibleState.currentState && !visibleState.targetState) {
                    mounted.remove(idx)
                }
                AnimatedVisibility(
                    visibleState = visibleState,
                    enter = fadeIn(tween(220)) + slideInHorizontally(tween(220)) { it / 3 },
                    exit = fadeOut(tween(180)),
                    modifier = Modifier.zIndex((TOOL_STACK_MAX_VISIBLE - pos).toFloat()).offset(x = x),
                ) {
                    AgentPickMiniTicketSkeleton(
                        modifier = Modifier.shadow(
                            elevation = 9.dp,
                            shape = RoundedCornerShape(AgentTicketGeometry.MINI_CORNER),
                            ambientColor = Color.Black.copy(alpha = 0.55f),
                            spotColor = Color.Black.copy(alpha = 0.55f),
                        ),
                    )
                }
            }
        }
    }
}

/** How many scaffolds render at once before the oldest fall off the front. */
internal const val TOOL_STACK_MAX_VISIBLE = 7

/** Comfortable peek while the deck is still growing. */
internal val TOOL_STACK_IDEAL_PEEK = 96.dp

/** Never overlap tighter than this, so every ticket keeps a readable sliver. */
internal val TOOL_STACK_MIN_PEEK = 28.dp

/**
 * The window of ticket indices currently on screen: the newest [TOOL_STACK_MAX_VISIBLE]
 * of [count]. Empty when nothing has been dealt yet.
 */
internal fun toolActivityVisibleIndices(count: Int, maxVisible: Int = TOOL_STACK_MAX_VISIBLE): List<Int> {
    if (count <= 0) return emptyList()
    val lower = maxOf(0, count - maxVisible)
    return (lower until count).toList()
}

/**
 * Horizontal step between fanned tickets. Grows at [TOOL_STACK_IDEAL_PEEK] until the last
 * ticket's trailing edge would leave the container, then clamps to exactly fill
 * the available travel, with [TOOL_STACK_MIN_PEEK] as the floor.
 */
internal fun toolActivityPeek(
    visibleCount: Int,
    containerWidth: Dp,
    cardWidth: Dp,
    idealPeek: Dp = TOOL_STACK_IDEAL_PEEK,
    minPeek: Dp = TOOL_STACK_MIN_PEEK,
): Dp {
    if (visibleCount <= 1) return 0.dp
    val travel = (containerWidth - cardWidth).coerceAtLeast(0.dp)
    val fillPeek = travel / (visibleCount - 1)
    return maxOf(minPeek, minOf(idealPeek, fillPeek))
}
